internal/deploy/yaml.go

```diff
@@ -0,0 +1,213 @@
+		var afterHooks []string
+		ketchYamlProcessConfig := make(map[string]ketchv1.KetchYamlProcessConfig)
+		for _, process := range *application.Processes {
+			processes = append(processes, ketchv1.ProcessSpec{
+				Name:  process.Name,
+				Cmd:   strings.Split(process.Cmd, " "),
+				Units: process.Units,
+				Env:   envs,
+			})
+			if process.Hooks.Restart.Before != "" {
+				beforeHooks = append(beforeHooks, process.Hooks.Restart.Before)
+			}
+			if process.Hooks.Restart.After != "" {
+				afterHooks = append(afterHooks, process.Hooks.Restart.After)
+			}
+
+			var ports []ketchv1.KetchYamlProcessPortConfig
+			for _, port := range process.Ports {
+				ports = append(ports, ketchv1.KetchYamlProcessPortConfig{
+					Protocol:   port.Protocol,
+					Port:       port.Port,
+					TargetPort: port.TargetPort,
+				})
+			}
+			if len(process.Ports) > 0 {
+				ketchYamlProcessConfig[process.Name] = ketchv1.KetchYamlProcessConfig{
+					Ports: ports,
+				}
+			}
+		}
+
+		// assign hooks and ports (kubernetes processConfig) to ketch yaml data
+		// NOTE: there is a disparity in that the yaml file format implies that hooks and ports
+		// are per-process, while the AppSpec makes them per-deployment.
+		ketchYamlData = ketchv1.KetchYamlData{
+			Hooks: &ketchv1.KetchYamlHooks{
+				Restart: ketchv1.KetchYamlRestartHooks{
+					Before: beforeHooks,
+					After:  afterHooks,
+				},
+			},
+			Kubernetes: &ketchv1.KetchYamlKubernetesConfig{
+				Processes: ketchYamlProcessConfig,
+			},
+		}
+	}
+	c := &ChangeSet{
+		appName:              *application.Name,
+		appVersion:           application.Version,
+		appType:              application.Type,
+		image:                application.Image,
+		description:          application.Description,
+		envs:                 application.Environment,
+		framework:            application.Framework,
+		dockerRegistrySecret: application.RegistrySecret,
+		builder:              application.Builder,
+		buildPacks:           application.BuildPacks,
+		appUnit:              application.AppUnit,
+		timeout:              &o.Timeout,
+		wait:                 &o.Wait,
+	}
+	if application.CName != nil {
+		c.cname = &ketchv1.CnameList{application.CName.DNSName}
+	}
+	if len(processes) > 0 {
+		c.processes = &processes
+		c.ketchYamlData = &ketchYamlData
+	}
+	c.applyDefaults()
+	return c, c.validate()
+}
+
+// apply defaults sets default values for a ChangeSet
+func (c *ChangeSet) applyDefaults() {
+	if c.appVersion == nil {
+		c.appVersion = &defaultVersion
+	}
+	if c.appType == nil {
+		c.appType = &typeApplication
+	}
+	c.yamlStrictDecoding = true
+	// building from source in PWD
+	if c.builder != nil && c.sourcePath == nil {
+		sourcePath := "."
+		c.sourcePath = &sourcePath
+	}
+	// default to AppUnits if process.Units is unset
+	if c.appUnit == nil {
+		c.appUnit = &defaultAppUnit
+	}
+	if c.processes != nil {
+		for i := range *c.processes {
+			if (*c.processes)[i].Units == nil {
+				if c.appUnit != nil {
```

Comment:
given that  `c.appUnit` is always set
```golang
if c.appUnit == nil {
		c.appUnit = &defaultAppUnit
}
```
can we simplify it a bit?
```golang
if c.processes != nil {
	for i := range *c.processes {
		if (*c.processes)[i].Units == nil {
			(*c.processes)[i].Units = c.appUnit
		}
	}
}


```

Reply:
Nice catch. Removed duplicated logic.